src/main/java/com/snowflake/kafka/connector/internal/telemetry/SnowflakeTelemetryPipeStatus.java

```diff
@@ -338,4 +370,91 @@ private void registerPipeJMXMetrics(
       LOGGER.warn("Metrics already present:{}", ex.getMessage());
     }
   }
+
+  // --------------- Setter for Offset counts --------------- //
+
+  public void setProcessedOffset(long processedOffset) {
+    this.processedOffset.set(processedOffset);
+  }
+
+  public void setFlushedOffset(long flushedOffset) {
+    this.flushedOffset.set(flushedOffset);
+  }
+
+  public void setCommittedOffset(long committedOffset) {
+    this.committedOffset.set(committedOffset);
+  }
+
+  /**
+   * Either keeps the same offset or updates the purgedOffset if a higher value offset is found in
+   * insertReport Snowpipe API
+   *
+   * @param unaryOperator the function to apply on purgedOffset
+   */
+  public void setPurgedOffsetAtomically(LongUnaryOperator unaryOperator) {
+    this.purgedOffset.updateAndGet(unaryOperator);
+  }
+
+  // --------------- File Counts at various stages of ingestion --------------- //
+
+  public void addAndGetFileCountOnStage(long fileCountOnStage) {
+    this.fileCountOnStage.addAndGet(fileCountOnStage);
+  }
+
+  public void addAndGetFileCountOnIngestion(long fileCountOnIngestion) {
+    this.fileCountOnIngestion.addAndGet(fileCountOnIngestion);
+  }
+
+  public void addAndGetFileCountPurged(long fileCountPurged) {
+    this.fileCountPurged.addAndGet(fileCountPurged);
+  }
+
+  public long incrementAndGetCleanerRestartCount() {
+    return this.cleanerRestartCount.incrementAndGet();
+  }
+
+  public void addAndGetTotalNumberOfRecord(long totalNumberOfRecord) {
+    this.totalNumberOfRecord.addAndGet(totalNumberOfRecord);
+  }
+
+  public void addAndGetTotalSizeOfData(long totalSizeOfData) {
+    this.totalSizeOfData.addAndGet(totalSizeOfData);
+  }
+
+  public void addAndGetMemoryUsage(long memoryUsage) {
+    this.memoryUsage.addAndGet(memoryUsage);
+  }
+
+  public void resetMemoryUsage() {
+    this.memoryUsage.set(0l);
+  }
+
+  // --------------- For testing --------------- //
```

Comment:
Should we add @VisibleForTesting for all the functions below?

Reply:
Good idea. Will do!